ChaCha20-Poly1305 for Armv8 (AArch64)

nebeid opened the pull request:
### Issues:
Resolves CryptoAlg-784

### Description of changes: 
This PR integrates the CL Vlad Krasnov opened for BoringSSL  [44364: Optimized ChaCha20-Poly1305 for AArch64](https://boringssl-review.googlesource.com/c/boringssl/+/44364). The CL was thoroughly reviewed by David Benjamin but not merged due to some outstanding comments which this PR addresses:
- The flag check when doing the final reduction in poly1305 was changed from `eq` to `cs`
- The CFI prologues and epilogues of open/seal were modified as recommended by David.
- Added Pointer Authentication instruction to the functions that are exported from this file based on [the explanation in this commit](https://github.com/awslabs/aws-lc/commit/fe049e4d186f253140e3d4a2570bba93386b75d8) as pointed out by David.

The first commit is the CL of Vlad without any changes.

### Call-outs:
- Finding a test vector that would testify that having `eq` as the condition checked in the final reduction as a bug is highly unlikely. This is explained below.

### Testing:
- The current tests against ChaCha20-Poly1305 continue to pass.
- More test vectors were produced using a Python script to try and investigate the `eq` bug. They passed as well. The required a final reduction, but didn't result in a the most significant word being non-zero after the reduction, which would have highlighted the bug.
- `objdump -W|Wf|WF` was used to confirm the value of the CFA and the locations of the registers relative to the CFA were as expected. See https://www.imperialviolet.org/2017/01/18/cfi.html.

#### Performance
```
Before:
 ./tool/bssl speed -filter "ChaCha20-Poly1305"
Did 1930000 ChaCha20-Poly1305 (16 bytes) seal operations in 1000289us (1929442.4 ops/sec): 30.9 MB/s
Did 862000 ChaCha20-Poly1305 (256 bytes) seal operations in 1001134us (861023.6 ops/sec): 220.4 MB/s
Did 212000 ChaCha20-Poly1305 (1350 bytes) seal operations in 1000315us (211933.2 ops/sec): 286.1 MB/s
Did 41000 ChaCha20-Poly1305 (8192 bytes) seal operations in 1019424us (40218.8 ops/sec): 329.5 MB/s
Did 21000 ChaCha20-Poly1305 (16384 bytes) seal operations in 1036779us (20255.0 ops/sec): 331.9 MB/s
After:
./tool/bssl speed -filter "ChaCha20-Poly1305"
Did 2705000 ChaCha20-Poly1305 (16 bytes) seal operations in 1000288us (2704221.2 ops/sec): 43.3 MB/s
Did 1423000 ChaCha20-Poly1305 (256 bytes) seal operations in 1000637us (1422094.1 ops/sec): 364.1 MB/s
Did 474000 ChaCha20-Poly1305 (1350 bytes) seal operations in 1001003us (473525.1 ops/sec): 639.3 MB/s
Did 98000 ChaCha20-Poly1305 (8192 bytes) seal operations in 1006024us (97413.2 ops/sec): 798.0 MB/s
Did 50000 ChaCha20-Poly1305 (16384 bytes) seal operations in 1004764us (49762.9 ops/sec): 815.3 MB/s
```
#### Explanation of the unlikeliness of finding a test vector
* the modulus in t2:t1:t0 = 3 : FF..FF : FF..FB, each being a 64 bit word; i.e. t2 = 3, t1 = all 1s.
* acc2 <= 4 after the previous reduction.
* It is highly likely to have borrow = 1 from acc1 - t1 since t1 is all FFs.
* So for almost all test vectors we have acc2 <= 4 and borrow = 1, thus (t2 = acc2 - t2 - borrow) will be 0 whenever acc > modulus. **It would be highly unlikely to find such a test vector with t2 > 0 after that final reduction:** Trying to craft that vector requires having acc and r of high values before their multiplication, yet ensuring that after the reduction (see Note) of their product, the resulting value of the accumulator has t2 = 4, all 1s in t1 and most of t0 so that no borrow occurs from acc1:acc0 - t1:t0.
* Note: the reduction is basically carried by folding over the top 64+62 bits once, then folding them again shifted left by 2, resulting in adding 5 times those bits.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dkostic commented:
So is the BTI and pointer authentication support, introduced in https://github.com/awslabs/aws-lc/commit/a0b49d63fdc33e54eac93674c86891d15d181d87, not needed for this implementation?

nebeid commented:
> So is the BTI and pointer authentication support, introduced in [a0b49d6](https://github.com/awslabs/aws-lc/commit/a0b49d63fdc33e54eac93674c86891d15d181d87), not needed for this implementation?

Thank you for pointing that out, Dusan. I added PAC instructions and explained why there was no need for BTI.
The tests pass but I am not certain the toolchain used in our CI may supports them yet.
